### ISSUE TITLE
Catch up subscriptions are difficult to use

### ISSUE THREAD

**Issue by alexrudd** [EventStore/EventStore-Client-Go#48]
Hey, I'm trying to create a catch-up subscription for a stream but I'm hitting both #33, and difficulties telling when the subscription has gone "live".

It seems like the tcp API has the [liveProcessingStarted](https://developers.eventstore.com/clients/dotnet/5.0/subscriptions/catchup-subscriptions.html#subscribing-to-a-stream) call back for this, but there's no equivalent for the gRPC API.

To get the functionality of being able to subscribe to a stream from the start and receive a call back when live, I had to write this:

```go
// SubscribeToStream subscribes from the start of a stream and calls the
// liveness call back when the latest event is reached.
func (sc *StoreClient) SubscribeToStream(
	ctx context.Context,
	stream string,
	eventHandler func(messages.RecordedEvent),
	livenessHandler func(streamrevision.StreamRevision),
) error {
	// Event store handlers.
	var (
		eventAppeared = func(re messages.RecordedEvent) { eventHandler(re) }
		subDropped    = func(reason string) { log.Printf("subscription to %s dropped", stream) } // TODO: handle async dropped subs
		liveOnce      = sync.Once{}
		goLiveOnce    = func(rev streamrevision.StreamRevision) { liveOnce.Do(func() { livenessHandler(rev) }) }
		streamIsEmpty = false
	)

	// If there's a liveness callback then find the position at which we will
	// go live.
	if livenessHandler != nil {
		res, err := sc.esc.ReadStreamEvents(ctx, direction.Backwards, stream, streamrevision.StreamRevisionEnd, 1, false)
		if err != nil && !errors.Is(err, eserrs.ErrStreamNotFound) {
			return fmt.Errorf("reading latest event: %w", err)
		}

		if len(res) == 1 {
			eventAppeared = func(re messages.RecordedEvent) {
				// Have we reached or passed the latest event?
				if re.EventNumber >= res[0].EventNumber {
					goLiveOnce(streamrevision.StreamRevision(re.EventNumber))
				}

				eventHandler(re)
			}
		} else {
			// There are no events in the stream so go live on func return.
			streamIsEmpty = true
			defer goLiveOnce(streamrevision.StreamRevision(streamrevision.StreamRevisionStart))
		}
	}

	if !streamIsEmpty {
		// Read the first event (this is not included in catch up subscriptions).
		// Attempt to read two event so we can tell if the stream size is greater
		// than one.
		res, err := sc.esc.ReadStreamEvents(ctx, direction.Forwards, stream, streamrevision.StreamRevisionStart, 2, false)
		if err != nil && !errors.Is(err, eserrs.ErrStreamNotFound) {
			return fmt.Errorf("reading first event: %w", err)
		}

		// We attempted to read two events but only received one, the stream must
		// only contain one event.
		if len(res) == 1 {
			eventHandler(res[0])

			// There was only one event is the stream, which we've passed to the
			// handler, so go live on func return.
			defer goLiveOnce(streamrevision.StreamRevision(res[0].EventNumber))
		} else {
			eventHandler(res[0])
		}
	}

	// Subscribe to stream.
	sub, err := sc.esc.SubscribeToStream(ctx, stream, streamrevision.StreamRevisionStart, true, eventAppeared, nil, subDropped)
	if err != nil {
		return fmt.Errorf("creating subscription: %w", err)
	}

	// Start the subscription.
	if err := sub.Start(); err != nil {
		return fmt.Errorf("starting subscription: %w", err)
	}

	return nil
}
```

It seems like the server has a concept of "going live", it would be great if this could be exposed in the gRPC API, possibly as just another content type in the ReadResp: https://github.com/EventStore/EventStore/blob/master/src/Protos/Grpc/streams.proto#L85

```
esdb-node      | [    1,47,16:07:14.600,INF] Catch-up subscription 1a4ebd24-f9bc-4020-ad69-c4e36e0aec96 to "my-stream"@"1" running...
esdb-node      | [    1,47,16:07:14.600,INF] Live subscription 1a4ebd24-f9bc-4020-ad69-c4e36e0aec96 to "my-stream" running from "1"...
```

This seems like a very common usage pattern but it's very difficult to implement and easy to do incorrectly and introduce bugs (I'm not 100% the above works for all cases).

**Comment by alexrudd** [comment]
Yeah my above code example doesn't work for the case where the stream doesn't exist at the time of subscription. The first event that is appended to the stream after the subscription starts will be ignored. I'm not sure how to work around this other than with polling in the case where the stream starts empty and only starting the subscription after the first event is appended.

Is there no way to subscribe inclusively from the start of a stream?

**Comment by YoEight** [comment]
This is fixed but only if you use at least 20.10.*
